cmd/object-handlers.go

```
@@ -450,8 +450,14 @@ func (api objectAPIHandlers) getObjectHandler(ctx context.Context, objectAPI Obj
 	// Automatically remove the object/version is an expiry lifecycle rule can be applied
 	if lc, err := globalLifecycleSys.Get(bucket); err == nil {
 		action := evalActionFromLifecycle(ctx, *lc, objInfo, false)
-		if action == lifecycle.DeleteAction || action == lifecycle.DeleteVersionAction {
-			globalExpiryState.queueExpiryTask(objInfo, action == lifecycle.DeleteVersionAction)
+		var success bool
+		switch action {
+		case lifecycle.DeleteVersionAction, lifecycle.DeleteAction:
+			success = applyExpiryRule(objInfo, false, action == lifecycle.DeleteVersionAction)
+		case lifecycle.DeleteRestoredAction, lifecycle.DeleteRestoredVersionAction:
+			success = applyExpiryRule(objInfo, true, action == lifecycle.DeleteRestoredVersionAction)
+		}
```

This is a bug that is fixed in this PR @krisis @vadmeste

---

cleaning up restored version successfully does not mean we should return an error on HEAD/GET. The remote version is still present - this clean up is best avoided here

---

That is a good catch but I would still trigger this here if necessary just not return an error to the client.